typing/typecore.ml

```diff
@@ -2778,6 +2804,18 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 mode_fun sargs ret
       when sargs <> [] && is_commu_ok com ->
         let lv = get_level ty_fun' in
         let (l, mode_arg, mode_ret) = ad in
+        begin match sargs with
+        | [] -> assert false
+        | (_lbl, sarg) :: rest ->
+          (* This is an obscure error, which often triggers on ill-typed applications.
+             Since it is relatively confusing, we want to delay issuing it until we're
+             sure the application is otherwise well-typed, because in ill-typed cases
+             the standard error is easier to understand *)
+          delayed_checks :=
+            (fun () -> check_local_application_complete ~env ~app_loc ~sarg ~rest
+                         ~mode_fun ~mode_arg ~mode_ret)
+            :: !delayed_checks
+        end;
```

Comment:
I think it would be better to avoid mixing this check up with the complicated argument gathering logic. The result of `collect_apply_args` already has most of the needed information for the check. I think it is just missing `mode_fun` in the `Arg (Known_arg ...)` and `Arg (Unknown_arg ...)` cases, and it's readily available at the point where those constructors are created. So rather than a higher-order ref, you can just iterate over that data at the point where you want to do the check (it's currently shadowed by the output of `type_apply_arg` but that's easy to fix).